lib/src/sse_handler.dart

```diff
@@ -0,0 +1,214 @@
+    'Access-Control-Allow-Credentials: true\r\n'
+    'Access-Control-Allow-Origin: $origin\r\n'
+    '\r\n\r\n';
+
+/// A bi-directional SSE connection between server and browser.
+class SseConnection extends StreamChannelMixin<String> {
+  /// Incoming messages from the Browser client.
+  final _incomingController = StreamController<String>();
+
+  /// Outgoing messages to the Browser client.
+  final _outgoingController = StreamController<String>();
+
+  Sink _sink;
+
+  /// How long to wait after a connection drops before considering it closed.
+  final Duration _keepAlive;
+
+  /// A timer counting down the KeepAlive period (null if connected).
+  Timer _keepAliveTimer;
+
+  /// The subscription that passes outgoing messages to the sink.
+  ///
+  /// This will be paused during the keepalive period and resumed upon reconnection.
+  StreamSubscription _outgoingStreamSubscription;
+
+  final _closedCompleter = Completer<void>();
+
+  /// Creates an [SseConnection] for the supplied [_sink].
+  ///
+  /// If [keepAlive] is supplied, the connection will remain active for this
+  /// period after a disconnect and can be reconnected transparently. If there
+  /// is no reconnect within that period, the connection will be closed normally.
+  ///
+  /// If [keepAlive] is not supplied, the connection will be closed immediately
+  /// after a disconnect.
+  SseConnection(this._sink, {Duration keepAlive}) : _keepAlive = keepAlive {
+    _outgoingStreamSubscription = _outgoingController.stream.listen((data) {
+      if (!_closedCompleter.isCompleted) {
+        try {
+          // JSON encode the message to escape new lines.
+          _sink.add('data: ${json.encode(data)}\n');
+          _sink.add('\n');
+        } catch (StateError) {
+          if (_keepAlive == null) {
+            rethrow;
+          }
+          // If we got here then the sink may have closed but the stream.onDone
+          // hasn't fired yet, so pause the subscription, re-queue the message
+          // and handle the error as a disconnect.
+          _handleDisconnect();
+          _outgoingController.add(data);
+        }
+      }
+    });
+    _outgoingController.onCancel = _close;
+    _incomingController.onCancel = _close;
+  }
+
+  /// The message added to the sink has to be JSON encodable.
+  @override
+  StreamSink<String> get sink => _outgoingController.sink;
+
+  // Add messages to this [StreamSink] to send them to the server.
+  /// [Stream] of messages sent from the server to this client.
+  ///
+  /// A message is a decoded JSON object.
+  @override
+  Stream<String> get stream => _incomingController.stream;
+
+  void _acceptReconnection(Sink sink) {
+    _keepAliveTimer.cancel();
+    _keepAliveTimer = null;
+    _sink = sink;
+    _outgoingStreamSubscription.resume();
+  }
+
+  void _handleDisconnect() {
+    if (_keepAlive == null) {
+      // Close immediately if we're not keeping alive.
+      _close();
+    } else if (_keepAliveTimer == null) {
```

Comment:
This can just be `!_keepAliveTimer?.isActive ?? true`

Reply:
This triggered a lint:

```
The value of the '?.' operator can be 'null', which isn't appropriate as an operand of a logical operator
```

Since we also checked this in `SseConnection`, I made a `bool get isInKeepAlivePeriod => _keepAliveTimer?.isActive ?? false;` to simplify the code a little.